File: java/leetcode/binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        return traverse(root, new ArrayList<Integer>());
    }
    
    public List<Integer> traverse(TreeNode root, List<Integer> result) {
        if (root != null) {
            result.add(root.val);
            result = traverse(root.left, result);
            result = traverse(root.right, result);
        }
        return result;
    }
}

File: python/leetcode/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        return self.traverse(root, [])
    
    def traverse(self, root, result):
        if (root is not None):
            result.append(root.val)
            result = self.traverse(root.left, result)
            result = self.traverse(root.right, result)
        return result
